File: boat-sample/src/test/java/sample/xyz/srclab/common/reflect/ReflectSample.java
package sample.xyz.srclab.common.reflect;

import org.testng.annotations.Test;
import xyz.srclab.common.reflect.Reflects;
import xyz.srclab.common.reflect.Types;
import xyz.srclab.common.test.TestLogger;

import java.lang.reflect.GenericArrayType;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.util.List;

public class ReflectSample {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testReflects() {
        Method method = Reflects.method(Object.class, "toString");
        String s = Reflects.invoke(method, new Object());
        //java.lang.Object@97c879e
        logger.log("s: {}", s);
    }

    @Test
    public void testTypes() {
        ParameterizedType type = Types.parameterizedType(List.class, String.class);
        GenericArrayType arrayType = Types.genericArrayType(type);
        //java.util.List<java.lang.String>[]
        logger.log("arrayType: {}", arrayType);
    }
}

File: boat-sample/src/test/kotlin/sample/kotlin/xyz/srclab/core/reflect/ReflectSample.kt
package sample.kotlin.xyz.srclab.core.reflect

import org.testng.annotations.Test
import xyz.srclab.common.reflect.genericArrayType
import xyz.srclab.common.reflect.invoke
import xyz.srclab.common.reflect.method
import xyz.srclab.common.reflect.parameterizedType
import xyz.srclab.common.test.TestLogger

class ReflectSample {

    @Test
    fun testReflects() {
        val method = Any::class.java.method("toString")
        val s = method.invoke<String>(Any())
        //java.lang.Object@97c879e
        logger.log("s: {}", s)
    }

    @Test
    fun testTypes() {
        val type = parameterizedType(MutableList::class.java, String::class.java)
        val arrayType = type.genericArrayType()
        //java.util.List<java.lang.String>[]
        logger.log("arrayType: {}", arrayType)
    }

    companion object {
        private val logger = TestLogger.DEFAULT
    }
}